File: library/src/main/java/com/yy/codex/uikit/UIViewHelpers.java
package com.yy.codex.uikit;

import android.support.annotation.NonNull;
import android.view.MotionEvent;

/**
 * Created by cuiminghui on 2017/1/16.
 */

class UIViewHelpers {

    static public UIView hitTest(@NonNull UIView view, @NonNull CGPoint point, @NonNull MotionEvent event) {
        UIView[] views = view.getSubviews();
        if (!view.isUserInteractionEnabled() && !(view.getAlpha() > 0)) {
            return null;
        }
        if (pointInside(view, point)) {
            for (UIView subview: views) {
                CGPoint convertedPoint = view.convertPoint(point, subview);
                UIView hitTestView = subview.hitTest(convertedPoint, event);
                if (hitTestView != null) {
                    return hitTestView;
                }
            }
            return view;
        }
        return null;
    }

    static public boolean pointInside(@NonNull UIView view, @NonNull CGPoint point) {
        double h = view.getFrame().size.height;
        double w = view.getFrame().size.width;
        double touchX = point.x;
        double touchY = point.y;
        if (touchY <= h && touchX <= w && touchY >= 0 && touchX >= 0) {
            return true;
        }
        return false;
    }

    @NonNull
    static public CGPoint convertPoint(@NonNull UIView view, @NonNull CGPoint point, @NonNull UIView toView) {
        if (view == toView) {
            return point;
        }
        CGPoint convertPoint;
        UIView toViewSuperView = toView;
        UIView superView = view;
        do {
            convertPoint = convertPointToSubView(view, point, toViewSuperView);
            toViewSuperView = toViewSuperView.getSuperview();
        } while (toViewSuperView != view && toViewSuperView != null);
        if (toViewSuperView == null) {
            do {
                convertPoint = convertPointToSubView(view, point, superView);
                superView = superView.getSuperview();
            } while (superView != toViewSuperView && superView != null);
        }
        if (superView == null) {
            toViewSuperView = toView;
            superView = view;
            do {
                UIView innerToViewSuperView = toViewSuperView.getSuperview();
                UIView innerSuperView = superView.getSuperview();
                if (innerToViewSuperView == superView) {
                    break;
                }
                convertPoint = convertPointToSuperView(view, convertPoint, superView);
                if (innerToViewSuperView != null) {
                    toViewSuperView = innerToViewSuperView;
                }
                if (innerSuperView != null) {
                    superView = innerSuperView;
                }
            } while (toViewSuperView != superView);
            if (toViewSuperView != null && superView != null) {
                double toX = toView.getFrame().origin.x;
                double toY = toView.getFrame().origin.y;
                return new CGPoint(convertPoint.x - toX, convertPoint.y - toY);
            }
        }
        return convertPoint;
    }

    @NonNull
    static private CGPoint convertPointToSuperView(@NonNull UIView view, @NonNull CGPoint point, @NonNull UIView superView) {
        double x = superView.getFrame().origin.x;
        double y = superView.getFrame().origin.y;
        return new CGPoint(point.x + x, point.y + y);
    }

    @NonNull
    static private CGPoint convertPointToSubView(@NonNull UIView view, @NonNull CGPoint point, @NonNull UIView subView) {
        double x = subView.getFrame().origin.x;
        double y = subView.getFrame().origin.y;
        return new CGPoint(point.x - x, point.y - y);
    }

}

File: library/src/main/java/com/yy/codex/uikit/UIViewHelpers.kt
package com.yy.codex.uikit

import android.view.MotionEvent

/**
 * Created by cuiminghui on 2017/1/16.
 */

internal object UIViewHelpers {

    fun hitTest(view: UIView, point: CGPoint, event: MotionEvent): UIView? {
        val views = view.subviews.reversed()
        if (!view.userInteractionEnabled && view.alpha <= 0) {
            return null
        }
        if (pointInside(view, point)) {
            for (subview in views) {
                if (!subview.userInteractionEnabled || subview.alpha <= 0) {
                    continue
                }
                val convertedPoint = view.convertPoint(point, subview)
                val hitTestView = subview.hitTest(convertedPoint, event)
                if (hitTestView != null) {
                    return hitTestView
                }
            }
            return view
        }
        return null
    }

    fun pointInside(view: UIView, point: CGPoint, allowMovement: CGPoint? = null): Boolean {
        val h = view.frame.size.height
        val w = view.frame.size.width
        val touchX = point.x
        val touchY = point.y
        if (touchY <= h + (allowMovement?.y ?: 0.0) && touchX <= w + (allowMovement?.x ?: 0.0) && touchY >= -(allowMovement?.y ?: 0.0) && touchX >= -(allowMovement?.x ?: 0.0)) {
            return true
        }
        return false
    }

    fun convertPoint(view: UIView, point: CGPoint, toView: UIView): CGPoint {
        if (view === toView) {
            return point
        }
        var viewRoot = view
        var viewX = viewRoot.frame.origin.x
        var viewY = viewRoot.frame.origin.y
        while (viewRoot.superview != null) {
            (viewRoot.superview as? UIScrollView)?.let {
                viewX -= it.contentOffset.x
                viewY -= it.contentOffset.y
            }
            viewRoot.superview?.let {
                viewRoot = it
            }
            viewX += viewRoot.frame.origin.x
            viewY += viewRoot.frame.origin.y
        }
        var toViewRoot = toView
        var toViewX = toViewRoot.frame.origin.x
        var toViewY = toViewRoot.frame.origin.y
        while (toViewRoot.superview != null) {
            (toViewRoot.superview as? UIScrollView)?.let {
                toViewX -= it.contentOffset.x
                toViewY -= it.contentOffset.y
            }
            toViewRoot.superview?.let {
                toViewRoot = it
            }
            toViewX += toViewRoot.frame.origin.x
            toViewY += toViewRoot.frame.origin.y
        }
        if (viewRoot !== toViewRoot) {
            return CGPoint(0.0, 0.0)
        }
        return CGPoint(viewX - toViewX + point.x, viewY - toViewY + point.y)
    }

    fun findRootView(view: UIView): UIView? {
        var rootView = view.superview
        while (rootView?.superview != null) {
            rootView = rootView?.superview
        }
        return rootView
    }

}
